File: 백준/Gold/27163. 벚꽃 내리는 시대에 결투를/벚꽃 내리는 시대에 결투를.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] line = br.readLine().split(" ");
        int n = Integer.parseInt(line[0]);
        int a = Integer.parseInt(line[1]);
        int l = Integer.parseInt(line[2]);
        int[][] dp = new int[n + 1][l + 1];
        int[][] lg = new int[n + 1][l + 1];
        int[][] ln = new int[n][2];
        for (int i = 0; i < n; i++) {
            line = br.readLine().split(" ");
            Arrays.fill(dp[i], -1);
            Arrays.fill(lg[i], -1);
            for (int j = 0; j < 2; j++) ln[i][j] = Integer.parseInt(line[j]);
        }
        Arrays.fill(dp[n], -1);
        Arrays.fill(lg[n], -1);
        dp[0][l] = a;
        for (int i = 0; i < n; i++) {
            int x = ln[i][0];
            int y = ln[i][1];
            for (int j = 1; j <= l; j++) {
                if (dp[i][j] == -1) continue;
                if (y >= 0 && j > y && dp[i + 1][j - y] < dp[i][j]) {
                    dp[i + 1][j - y] = dp[i][j];
                    lg[i + 1][j - y] = y;
                }
                if (x >= 0) {
                    int mx = dp[i][j] - x;
                    if (y < 0 && mx < 0) mx = 0;
                    if (dp[i + 1][j] < mx) {
                        dp[i + 1][j] = mx;
                        lg[i + 1][j] = 0;
                    }
                }
            }
        }
        List<Character> rst = new ArrayList<>();
        int q = 1;
        for (int i = 1; i <= l; i++) {
            if (lg[n][i] != -1) {
                System.out.println("YES");
                int k = i;
                for (int j = n; j > 0; j--) {
                    if (lg[j][k] != 0 || ln[j - 1][1] == 0) rst.add('L');
                    else rst.add('A');
                    k += lg[j][k];
                }
                for (int j = rst.size() - 1; j >= 0; j--) System.out.print(rst.get(j));
                q = 0;
                break;
            }
        } 
        if (q == 1) System.out.println("NO");
    }
}

File: 백준/Gold/27163. 벚꽃 내리는 시대에 결투를/벚꽃 내리는 시대에 결투를.py
import sys
input = sys.stdin.readline

n, a, l = map(int, input().split())
dp = [[-1] * (l + 1) for _ in range(n + 1)]
dp[0][l] = a
xy = [list(map(int, input().split())) for _ in range(n)]
for i in range(n):
    x, y = xy[i]
    for j in range(1, l + 1):
        if dp[i][j] == -1: continue
        if y >= 0 and j > y and dp[i + 1][j - y] < dp[i][j]:
            dp[i + 1][j - y] = dp[i][j]
        if x >= 0:
            mx = dp[i][j] - x
            if y < 0 and mx < 0: mx = 0
            if dp[i + 1][j] < mx:
                dp[i + 1][j] = mx
rst = []
for ll in range(1, l + 1):
    if dp[n][ll] != -1:
        print('YES')
        aa = dp[n][ll]
        for i in range(n, 0, -1):
            x, y = xy[i - 1]
            if x == -1: rst.append('L'); ll += y
            elif y == -1: rst.append('A'); aa += x
            elif ll + y <= l and dp[i - 1][ll + y] == aa: rst.append('L'); ll += y
            else: rst.append('A'); aa += x
        print(''.join(rst[::-1])); break
else: print('NO')